index.bs

```diff
@@ -2223,14 +2377,143 @@ interface MLGraph {
                     1. Set the values of |inputTensor| to the values of |value|.{{MLInput/resource}}.
                 1. If |value| is an {{MLResource}}, then:
                     1. Set the values of |inputTensor| to the values of |value|.
-                1. Set the input of |this|.{{MLGraph/[[implementation]]}} that is associated with |key| to |inputTensor|.
+                1. Set the input of |graph|.{{MLGraph/[[implementation]]}} that is associated with |key| to |inputTensor|.
+            1. For each |key| -> |value| of |outputs|:
+                1. Issue a compute request for output of |graph|.{{MLGraph/[[implementation]]}} that is associated with |key|.
+                1. Wait for the compute request to be completed.
+                1. If there is an error returned by |graph|.{{MLGraph/[[implementation]]}}, then:
+                    1. Throw an {{OperationError}} {{DOMException}} and stop.
+                1. Else:
+                    1. Let |outputTensor| be the output tensor returned by |graph|.{{MLGraph/[[implementation]]}}.
+                    1. If the kind of |value| is not compatible with the value type of |outputTensor|, then throw a {{DataError}} {{DOMException}} and stop.
+                    1. Let |outputSize| be 1.
+                    1. For each |dimension| of dimensions of |outputTensor|:
+                        1. Set |outputSize| to the product of |outputSize| and |dimension|.
+                    1. If |outputSize| is greater than the length of |value|, then:
+                        1. Throw a {{DataError}} {{DOMException}} and stop.
+                    1. Else:
+                        1. Set the values of |value| to the values of |outputTensor|.
+            1. Return Promise<{{MLNamedOutputs}}>.
+        </div>
+</dl>
+
+## MLCommandEncoder ## {#api-mlcommandencoder}
+The {{MLCommandEncoder}} interface represents a method of execution that synchronously records the computational workload of a compiled graph {{MLGraph}} to a GPU command buffer {{GPUCommandBuffer}} on the calling thread. Since the workload is not immediately executed, just recorded, this method allows more flexibility for the caller to determine how and when the recorded commands will be submitted for execution on the GPU relative to other GPU workload on the same queue. This method of execution requires an {{MLContext}} created with {{GPUDevice}}.
+
+<script type=idl>
+typedef (GPUBuffer or GPUTexture) MLGPUResource;
+
+dictionary MLGPUInput {
+  required MLGPUResource resource;
+  required sequence<long> dimensions;
+};
+
+typedef record<DOMString, (MLGPUResource or MLGPUInput)> MLNamedGPUInputs;
+typedef record<DOMString, MLGPUResource> MLNamedGPUOutputs;
+
+[SecureContext, Exposed=(Window, DedicatedWorker)]
+interface MLCommandEncoder {
+  constructor(MLContext context);
+
+  undefined initializeGraph(MLGraph graph, MLNamedGPUInputs inputs);
```

Comment:
Normally, an `initialize` method on an interface implies that the operation initializes the host interface object, which makes the host object stateful. I explicitly named this method `initializeGraph` because the operation would initialize the graph input param to the method, and not the host object itself. It makes the graph stateful, not the encoder.

Reply:
> Should MLExecution and MLAwaitedExecution expose initialize 

`MLExecution` is CPU only, and so there isn't a need to initialize and pre-process the weights i.e. graph constants and inputs are treated the same way.

`MLAwaitedExecution` when used for a GPU context, will need to initialize and pre-process the weights before the first inference call. But that can be done without defining the `initializeGraph` method because weights are simply constant operands bound to the graph at construction time. The implementation can already distinguish the weight constants from the rest of the named inputs, so initializing them could technically be done before the first inference call (i.e. `compute`).

The real reason that only `MLCommandEncoder` will need a separate `initializeGraph` method is that unlike `MLExecution` and `MLAwaitedExecution`, the command encoder does not actually carry out the execution of the compute workload. It merely records the workload onto the command buffer on behalf of the caller. This means that it is the caller's responsibility to track the weight constant tensors, and make sure that constant initialization workload is recorded separately by the `initializeGraph` call. 

Another way to think about it is that the command encoder is a lower level API than the execution API. And, while the execution API can "cache" the constant's initialization result before making the first inference call to compute the output result, the encoder API cannot, it can only record the act of initialization, as a workload separated from the inference workload captured by the subsequent `dispatch` call.

Reply:
A CPU backend may also need to pre-process the weights, e.g., conversion to a SIMD instruction optimized block memory layout. And as the discussion [above](https://github.com/webmachinelearning/webnn/pull/255#discussion_r816500166), I would suggest `MLExecution` supports both CPU and GPU. 

The graph initialization, e.g., constants pre-processing, may be time consuming. Doing it in sync `MLGraphBuilder::build` may block the main thread. Hosting a sync version in `MLExecution` and an async version in `MLAwaitedExecution` may be a solution.

Regarding to the resource type, the `initialize` of `MLExecution` and `MLAwaitedExecution` could only accept `ArrayBufferView`, while `initializeGraph` of `MLCommandEncorder` only accepts `GPUBuffer`. This would align with the execution methods respectively (with my suggestion in [comment](https://github.com/webmachinelearning/webnn/pull/255#discussion_r816489483)) and simplify the usage.

In this way, the graph building code could be shared for different execution methods. Otherwise, the callers need to call `MLGraphBuilder::constant` to bind constant to a graph for `MLExecution` and `MLAwaitedExecution`, but call `MLGraphBuilder::input` and bind the constant in `MLCommandEncoder::initializeGraph`. That leads to different graph building code.